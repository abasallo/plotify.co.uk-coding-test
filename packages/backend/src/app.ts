import express, { Express } from 'express'
import cors from 'cors'
import bodyParser from 'body-parser'

import index from './express/routes/index'

import { initialiseApolloServer } from './graphql/init'

export const app: Express = express()

initialiseApolloServer(app)
  .then(() => console.log('Apollo Server initialised'))
  .catch((error) => console.error('Apollo Server initialisation error: ' + error))

app.use(cors())

app.use(bodyParser.urlencoded({ extended: false }))
app.use(bodyParser.json())

app.use('/', index)
