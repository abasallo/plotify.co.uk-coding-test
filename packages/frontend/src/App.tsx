import React from 'react'
import './App.css'

import { BrowserRouter as Router, Route, Routes } from 'react-router-dom'

import { MuiThemeProvider, StylesProvider } from '@material-ui/core/styles'
import CssBaseline from '@material-ui/core/CssBaseline'

import { customMUITheme } from './customMUITheme'

import Home from './views/home/Home'
import Main from './views/main/Main'
import RandomFact from './views/randomFact/RandomFact'

import constants from './modules/constants'

const App = () => (
  <MuiThemeProvider theme={customMUITheme}>
    <StylesProvider injectFirst>
      <CssBaseline />
      <Router>
        <Routes>
          <Route path={constants.ROUTE_URL_ROOT} element={<Home />} />
          <Route path={`${constants.ROUTE_URL_COMPANY}/${constants.ROUTE_URL_COMPANY_PARAMS}`} element={<Main />} />
          <Route path={constants.ROUTE_URL_RANDOM_FACT} element={<RandomFact />} />
        </Routes>
      </Router>
    </StylesProvider>
  </MuiThemeProvider>
)

export default App
