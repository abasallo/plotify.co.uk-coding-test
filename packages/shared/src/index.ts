export * from './model/company.model'
export * from './model/phase.model'
export * from './model/task.model'
