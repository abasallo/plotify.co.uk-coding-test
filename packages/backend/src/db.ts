import { AppOrmModel } from './orm/model/app.model'
import { initialiseSequelize } from './orm/initialiseSequelize'

import { initialiseTestDatabase } from './modules/initialiseTestDatabase'

export const model: Promise<AppOrmModel> = process.env.CONNECTION_URL
  ? initialiseSequelize(process.env.CONNECTION_URL)
  : initialiseTestDatabase()
