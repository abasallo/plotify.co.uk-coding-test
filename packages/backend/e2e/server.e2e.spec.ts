import * as http from 'http'

import axios from 'axios'

import { ReasonPhrases, StatusCodes } from 'http-status-codes'

import { app } from '../src/app'

const TEST_SERVER_PORT = 4001

let server: http.Server

beforeAll(async () => {
  server = app.listen({ port: TEST_SERVER_PORT }, () => console.log(`Test Server initialised on port: ${TEST_SERVER_PORT}`))
})

afterAll(() => {
  server.close()
})

describe('Server E2E tests', () => {
  describe('Health Check E2E tests', () => {
    test('Expected response for Health Check', async () => {
      const { status, statusText, data } = await axios({
        method: 'get',
        url: `http://localhost:${TEST_SERVER_PORT}/`
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })

  describe('Company E2E tests', () => {
    test('GetCompanies', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/graphql/`,
        data: {
          query: `
            query GetCompanies {
              getCompanies {
                id,
                name
              }
            }
            `
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })

  describe('Phase E2E tests', () => {
    test('GetPhases', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/graphql/`,
        data: {
          query: `
              query GetPhases($companyId: String!) {
                getPhases(companyId: $companyId) {
                  id,
                  name
                }
              }
            `,
          variables: {
            companyId: '1'
          }
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })
  })

  describe('Task E2E tests', () => {
    test('GetTasks', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/graphql/`,
        data: {
          query: `
              query GetTasks($phaseId: String!) {
                getTasks(phaseId: $phaseId) {
                  id,
                  text,
                  completed
                }
              }
            `,
          variables: {
            phaseId: '1'
          }
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot()
    })

    test('ToggleTaskCompletion', async () => {
      const { status, statusText, data } = await axios({
        method: 'post',
        url: `http://localhost:${TEST_SERVER_PORT}/graphql/`,
        data: {
          query: `
            mutation ToggleTaskCompletion($id: ID!) {
              toggleTaskCompletion(id: $id) {
                id,
                text,
                completed
              }
            }
          `,
          variables: {
            id: 1
          }
        }
      })

      expect(status).toEqual(StatusCodes.OK)
      expect(statusText).toEqual(ReasonPhrases.OK)

      expect(data).toMatchSnapshot({
        data: {
          toggleTaskCompletion: {
            completed: expect.any(Boolean)
          }
        }
      })
    })
  })
})
