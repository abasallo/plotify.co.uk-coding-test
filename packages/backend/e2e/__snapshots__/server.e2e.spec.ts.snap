// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server E2E tests Company E2E tests GetCompanies 1`] = `
{
  "data": {
    "getCompanies": [
      {
        "id": "1",
        "name": "Company I",
      },
      {
        "id": "2",
        "name": "Company II",
      },
      {
        "id": "3",
        "name": "Company III",
      },
    ],
  },
}
`;

exports[`Server E2E tests Health Check E2E tests Expected response for Health Check 1`] = `"OK"`;

exports[`Server E2E tests Phase E2E tests GetPhases 1`] = `
{
  "data": {
    "getPhases": [
      {
        "id": "1",
        "name": "Phase I",
      },
      {
        "id": "2",
        "name": "Phase II",
      },
      {
        "id": "3",
        "name": "Phase III",
      },
    ],
  },
}
`;

exports[`Server E2E tests Task E2E tests GetTasks 1`] = `
{
  "data": {
    "getTasks": [
      {
        "completed": false,
        "id": "1",
        "text": "Task I",
      },
    ],
  },
}
`;

exports[`Server E2E tests Task E2E tests ToggleTaskCompletion 1`] = `
{
  "data": {
    "toggleTaskCompletion": {
      "completed": Any<Boolean>,
      "id": "1",
      "text": "Task I",
    },
  },
}
`;
